ADBDEV-7691, ADBDEV-7690: Track PQresult allocations in server side, Gracefully handle OOM when cancelling a query

bandetto opened the pull request:
When querying `pg_locks` (or using `pg_lock_status()`), approximately 75% of backend memory allocations for resulting tuples weren't registered with Vmtracker or Resource Group Control. This memory would also leak if the query was cancelled or failed.

This happened because `CdbDispatchCommand()`, which is used by `pg_locks`, internally uses libpq to obtain the results, which were allocated as `PQresult` structures with bare `malloc()`, even on the server side.

Another issue is that if a query fails under resource group manager due to a lack of memory, the query is cancelled due to an error. The coordinator would still try to receive tuples from segments upon cancellation. Retrieving tuples requires more memory allocations, which may cause the same OOM error again. The query will get cancelled due to an error, entering into a recursion and leading to an eventual crash.

The first commit fixes both the memory leak and untracked memory issues by enforcing Vmtracker routines for `PGresult` allocations on the server-side. The original code is unchanged from #1231, besides addressing errcode-related macro redefinition warnings by un-defining them first.

The second commit prevents error recursion, upon query cancellation. It implements proper query cancellation by explicitly ignoring any libpq data from segments without extra memory allocations. The said issue cannot be reproduced without first commit. The original code was updated since #1231, based on #1668.

The PR contains several patches as separate, atomic commits, and therefore should be rebased. ADCC extension tests should pass before merging the PR.

---

When approved the PR will be moved to [Greengage](https://github.com/GreengageDB/greengage) repo for merge. ABI tests for `adb-6.x-dev` are broken.

RekGRpth commented on src/test/isolation2/input/resgroup/resgroup_oom.source:
```diff
@@ -0,0 +1,49 @@
+CREATE RESOURCE GROUP rg_oom_test
+WITH (cpu_rate_limit=20, memory_limit=20, memory_shared_quota=100);
+CREATE ROLE role_oom_test RESOURCE GROUP rg_oom_test;
+CREATE OR REPLACE FUNCTION gp_mock_cdbdispatchcommand(amount int)
```
```suggestion
CREATE FUNCTION gp_mock_cdbdispatchcommand(amount int)
```

bandetto replied:
I thinks that's unnecessary.

RekGRpth replied:
This will make sure that no one has left such a function created before.

bandetto replied:
For what?

RekGRpth commented on src/backend/cdb/dispatcher/cdbdisp_async.c:
```diff
@@ -914,38 +989,41 @@ static void
 
 	for (i = 0; i < pParms->dispatchCount; i++)
-	{
-		char		errbuf[256];
-		bool		sent = false;
-		CdbDispatchResult *dispatchResult = pParms->dispatchResultPtrArray[i];
+		signalQE(pParms->dispatchResultPtrArray[i], pParms->waitMode);
+}
 
-		Assert(dispatchResult != NULL);
-		SegmentDatabaseDescriptor *segdbDesc = dispatchResult->segdbDesc;
+/*
+ * Send finish or cancel signal to QE if needed.
+ */
+static void
+signalQE(CdbDispatchResult *dispatchResult, DispatchWaitMode waitMode)
```
What is the point now of refactoring the `signalQEs` function and adding a new `signalQE` function?

RekGRpth commented on src/backend/cdb/dispatcher/cdbdisp_async.c:
```diff
@@ -822,6 +825,60 @@ handlePollError(CdbDispatchCmdAsync *pParms)
+	pqSaveErrorResult(conn);
+
+	/* Make sure PQgetResult() calls are not blocking. */
+	PQconsumeInput(conn);
```
Why don't you check what this function returned?!

bandetto replied:
Why?

RekGRpth replied:
The function returns 0 on error, and you ignore it!

bandetto replied:
Why shouldn't it be ignored? Errors will be caught by `PQgetResult()` and a condition at the end.

RekGRpth replied:
see similar function `processResults`

bandetto replied:
I still can't understand you and I don't see a point in checking the return code. Can you please explain the reason?

RekGRpth commented on src/backend/cdb/dispatcher/cdbdisp_async.c:
```diff
@@ -822,6 +825,60 @@ handlePollError(CdbDispatchCmdAsync *pParms)
+	/* Free some memory and replace current result with a fatal error dummy. */
+	pqSaveErrorResult(conn);
+
+	/* Make sure PQgetResult() calls are not blocking. */
```
Actually, according to the documentation, to check if the `PQgetResult` function is blocked, you need to call the `PQisBusy` function.

bandetto replied:
https://www.postgresql.org/docs/current/libpq-async.html:

> Note that PQgetResult will block only if a command is active and the necessary response data has not yet been read by PQconsumeInput .

RekGRpth commented on src/test/isolation2/input/parallel_retrieve_cursor/fault_inject.source:
```diff
@@ -249,3 +249,25 @@ SELECT gp_inject_fault('fetch_tuples_from_endpoint', 'resume', dbid)
+
+-- Test 8: QD shouldn't hang waiting.
+
+DROP TABLE IF EXISTS t3;
```
Shouldn't we put this on the ignore block?

bandetto replied:
Nitpick, I thinks that's unnecessary work.

RekGRpth replied:
If some other test that was executed long before yours leaves a t3 table behind, your test will fail.

bandetto replied:
All of the other tests drop their tables `t1`, `t2` the same way. It's rather a task of refactoring.

RekGRpth replied:
For new code, you can do it right away.

RekGRpth commented on src/test/isolation2/input/parallel_retrieve_cursor/fault_inject.source:
```diff
@@ -249,3 +249,25 @@ SELECT gp_inject_fault('fetch_tuples_from_endpoint', 'resume', dbid)
 
 DROP TABLE t2;
+
+SELECT gp_inject_fault('alloc_endpoint_slot_full', 'reset', 2);
+SELECT gp_inject_fault('alloc_endpoint_slot_full_reset', 'reset', 2);
```
Before this patch, the test left these fault injectors unreset?!

bandetto replied:
Yeah.

bandetto committed:
Track PQresult allocations in server side (#1804)

When querying pg_locks (or using `pg_lock_status()`), approximately 75% of
backend memory allocations for resulting tuples weren't registered with
Vmtracker or Resource Group Control. This memory would also leak if the query
was cancelled or failed.

This happened because `CdbDispatchCommand()`, which was previously used by
`pg_locks`, called libpq to obtain the results that were allocated as PQresult
structures with bare `malloc()`, even on the server side.

This patch fixes both untracked memory issues by enforcing Vmtracker routines
for `PGresult` allocations on the server-side.

Including `postgres.h` in frontend code causes several errcode-related macro
redefinition warnings. They are now un-definined first.

Please note that the original error from ADBDEV-7145, which is an
`AbortTransaction()` recurstion due to an OOM error while cancelling a query,
is fixed only in the next commit.

Ticket: ADBDEV-7691

bandetto committed:
Gracefully handle OOM when cancelling a query (#1804)

Previosly, if a query fails under resource group manager due to a lack of
memory, the query is cancelled due to an error.

The coordinator would still try to receive tuples from segments upon
cancellation. Retrieving tuples requires more memory allocations, which may
cause the same OOM error again. The query will get cancelled due to an error,
entering into a recursion and leading to an eventual crash.

This patch implements proper query cancellation by explicitly ignoring any
libpq data from segments without extra memory allocations.

Ticket: ADBDEV-7690

RekGRpth commented:
Still, I suggest separating the patch for correcting the recursive error into a separate PR and accepting it first.

RekGRpth commented on src/backend/cdb/dispatcher/cdbdisp_async.c:
```diff
@@ -866,6 +922,24 @@ handlePollSuccess(CdbDispatchCmdAsync *pParms,
+			 * might not be enough memory to discard the result properly.
+			 * Let's get the big guns out.
+			 */
+			tryToWipeResults(dispatchResult);
```
Your `tryToWipeResults` function is very similar to the `processResults` function: in both functions
1) first the `PQconsumeInput` function is called
2) then there is a loop until `PQisBusy`
3) inside the loop the `PQgetResult` function is called
Maybe it would be easier to add the `pParms->waitMode == DISPATCH_WAIT_CANCEL` argument to the `processResults` function and its processing?

bandetto replied:
`processResults()` is complicated and I'll risk breaking something else when making changes there. 

`cdbconn_discardResults()` is also almost identical to the added function. But `tryToWipeResults()` was added for a reason and does only what it needs to.

bimboterminator1 replied:
Again, why we cannot use cdbconn_discardResults()? 
And I agree with @RekGRpth about avoiding injecting a special case this way. Are you sure there is not any approach that looks more organic?

bandetto replied:
> Again, why we cannot use cdbconn_discardResults()

Because it's used in other context and I'd prefer not to change existing behavior.

> Are you sure there is not any approach that looks more organic?

I think the current state of the patch is the most straightforward and safe way to make this change.

bimboterminator1 commented on src/interfaces/libpq/fe-protocol3.c:
```diff
@@ -529,7 +529,7 @@ pqParseInput3(PGconn *conn)
 					if (pqGetInt(&conn->result->extraslen, 4, conn))
 						return;
-					conn->result->extras = malloc(conn->result->extraslen);
+					conn->result->extras = pqPalloc(conn->result->extraslen);
```
why don't you replace any realloc calls? For example in fe-misc.c. PQEndcopy() uses it. Or the purpose is to focus only on result-related allocations?

bandetto replied:
Yes, the patch forces Vmtracker only for PQresults, since they make up for most of the allocations.

bimboterminator1 commented on src/backend/cdb/dispatcher/cdbdisp_async.c:
```diff
@@ -822,6 +822,60 @@ handlePollError(CdbDispatchCmdAsync *pParms)
+	if (PQisBusy(conn) && PQstatus(conn) != CONNECTION_BAD)
+	{
+		/* Some work is still remaining until we can die. */
+		return;
```
why we don't set `dispatchResult->stillRunning = false;` in this case?

bandetto replied:
Because the last part of the message haven't arrived yet and we we'll block waiting on next `PQgetResult()`.

bimboterminator1 reviewed (commented):
Why do you write about CdbDispatchCommand in pg_locks_status if it has been removed from there? Write description in way more consistent with current changes , please

bimboterminator1 reviewed (approved):
LGTM. Update PR description.

bandetto commented:
> I seem to have managed to reproduce the bug fixed by the second commit without using the first commit.

Injecting an error into the middle of a function does not really check the fundamental issue behind the task, which are failed memory allocations. Your solution checks only for re-entrance into `processResults()` and `handlePollSuccess()`.
